### PR TITLE
UPDATE ADD API key auth support to bugzilla

### PR DESCRIPTION
**Pull request by K-Gazdikova**
- Modified the code a bit to include different classes for auhentication
 - Added API key auth support for bugzilla tool operations.

### CODE REVIEW

File: ticketutil/bugzilla.py [hunk 1]
```diff
@@ -63,18 +115,24 @@ def _create_requests_session(self):
...
-            resp = r.json()
-            self.token = resp['token']
-            logging.debug("Create requests session: Status Code: {0}".format(r.status_code))
-            logging.info("Successfully authenticated to {0}.".format(self.ticketing_tool))
+            
+            if isinstance(self.auth, UPAuth):
+                r = s.get(self.auth_url, params=self.credentials, verify=False)
+                r.raise_for_status()
+                resp = r.json()
+                self.auth = TokenAuth(resp["token"])
```
Comment (dmranck) [inline ticketutil/bugzilla.py]: I'm a little confused with the difference between UPAuth and TokenAuth. Would a user ever use TokenAuth? Or would a user just use UPAuth, which generates a token? If the latter is true, then I would get rid of the separate TokenAuth class, and just have a function in UPAuth that generates this token.

File: ticketutil/bugzilla.py [hunk 2]
```diff
@@ -149,10 +207,12 @@ def _create_ticket_request(self, params):
...
         try:
-            if self.token:
-                params['token'] = self.token
+            #if self.token:
+            #    params['token'] = self.token
+            params.update(self.auth.getCredentials())
+            
 
-            r = self.s.post(self.rest_url, json=params)
+            r = self.s.post(self.rest_url, json=params, verify=False)
```
Comment (dmranck) [inline ticketutil/bugzilla.py]: Along the same lines as the previous comment about updating s.params, you can also set "s.verify = False" in _create_requests_session(), similar to how it's done in line 68 of ticket.py. This makes it so you don't need "verify=False" in every requests call throughout all the other methods.

Comment (K-Gazdikova) [inline ticketutil/bugzilla.py]: modified code for "s.verify = False" in _create_requests_session() and removed the remaining entries.

File: ticketutil/bugzilla.py [hunk 3]
```diff
@@ -149,10 +207,12 @@ def _create_ticket_request(self, params):
         """
         # Attempt to create ticket.
         try:
-            if self.token:
-                params['token'] = self.token
+            #if self.token:
+            #    params['token'] = self.token
+            params.update(self.auth.getCredentials())
```
Comment (K-Gazdikova) [inline ticketutil/bugzilla.py]: Made the suggested changes for _create_requests_session(), the code looks much cleaner 👍

File: ticketutil/bugzilla.py [hunk 4]
```diff
@@ -17,6 +15,58 @@
 else:
     logging.basicConfig(level=logging.INFO)
 
+class AUTHBase():
+    def getCredentials(self):
+        raise NotImplemented()
+    
+    
+
+class APIKeyAuth(AUTHBase):
```
Comment (K-Gazdikova) [inline ticketutil/bugzilla.py]: Updated

File: ticketutil/bugzilla.py [hunk 5]
```diff
@@ -17,6 +15,58 @@
...
+
+        self.apiKey = apiKey
+
+    def getKey(self):
+        return self.apiKey
+    
+    def getCredentials(self):
+        return {"api_key": self.getKey()}
+    
+class UPAuth(AUTHBase):
```
Comment (K-Gazdikova) [inline ticketutil/bugzilla.py]: Updated

File: ticketutil/bugzilla.py [hunk 6]
```diff
@@ -17,6 +15,58 @@
 else:
     logging.basicConfig(level=logging.INFO)
 
+class AUTHBase():
```
Comment (K-Gazdikova) [inline ticketutil/bugzilla.py]: Renamed

File: ticketutil/bugzilla.py [hunk 7]
```diff
@@ -197,12 +257,14 @@ def edit(self, **kwargs):
         kwargs = _prepare_ticket_fields(kwargs)
 
         params = kwargs
-        if self.token:
-            params['token'] = self.token
+        #if self.token:
+        #    params['token'] = self.token
+            
+        params.update(self.auth.getCredentials())
```
Comment (K-Gazdikova) [inline ticketutil/bugzilla.py]: Updated with the changes.

File: ticketutil/bugzilla.py [hunk 8]
```diff
@@ -322,8 +390,9 @@ def remove_cc(self, user):
         else:
             params = {'cc': {'remove': [user]}}
 
-        if self.token:
-            params['token'] = self.token
+        #if self.token:
+        #    params['token'] = self.token
+        params.update(self.auth.getCredentials())
```
Comment (K-Gazdikova) [inline ticketutil/bugzilla.py]: Updated

**Comment by dmranck** [comment]
Also, one additional comment - change all functions in the new auth classes to be all lower case with underscores separating the words. So, instead of getUsername, getPassword, getCredentials, name then get_username, get_password, get_credentials.

File: ticketutil/bugzilla.py [hunk 9]
```diff
@@ -31,63 +33,52 @@ def __init__(self, apiKey):
...
-    
+
 
 class BugzillaTicket(ticket.Ticket):
     """
     A BZ Ticket object. Contains BZ-specific methods for working with tickets.
     """
     def __init__(self, url, project, auth=None, ticket_id=None):
         self.ticketing_tool = 'Bugzilla'
+        self.resp = None
```
Comment (dmranck) [inline ticketutil/bugzilla.py]: I don't see this variable used anywhere else. Should it be deleted?

Comment (K-Gazdikova) [inline ticketutil/bugzilla.py]: removed  self.resp = None

**Comment by dmranck** [comment]
We're going to hold off on these changes for a bit. Instead of rolling out these new authentication classes only in Bugzilla, I'd like to make these changes across all tools at a later date (probably creating an auth.py file with our authentication classes). 

Closing this PR for now. @K-Gazdikova will take her authentication classes, put them in a new branch, and push to GitHub to be worked on at a later date.

In the meantime, Kumudini will push new code that adds api-key auth support to Bugzilla, consistent with our current way of authentication: auth={'apikey': <api_key>}

**Pull request closed without merging**